### ISSUE TITLE
API Questions

### ISSUE THREAD

**Issue by soreau** [skorokithakis/catt#204]
I like this project and was tinkering with making a gui for it but have a couple questions.

1) The cli cast() function supports casting files. The play_url() function in the api seems to only support urls and not files. Is there a way to play a file with the api?

2) Is there a way to get callbacks every second to update a seek bar and display time info or something similar for this purpose?

3) Is there a way to 'listen' for volume changes from other devices to update a gui knob in real time? I notice there's no info() function exposed in the api.

**Comment by theychx** [comment]
1: No, at this time it is not possible to play a file with the api.

2 & 3: `pychromecast` has such functionality, but it is not (yet) exposed in `catt.api`.

Examples using `pychromecast`:
```
class MediaListener:
    def new_media_status(self, status):
        # This mandatory method is called with new status whenever there is a change
        print(status)

class StatusListener:
    def new_cast_status(self, status):
        # This mandatory method is called with new status whenever there is a change
        print(status)

cast = pychromecast.Chromecast(<IPaddress of Chromecast>)
cast.wait()
media_listener = MediaListener()
cast.media_controller.register_status_listener(media_listener)
status_listener = StatusListener()
cast.register_status_listener(status_listener)
```

**Comment by theychx** [comment]
Some time soon, I plan on making a tui for `catt`, using `urwid`. That work might also add some kind of status callbacks to `catt.api`. I really should get started here in August :smiley: .

**Comment by soreau** [comment]
@theychx Thanks for the code, it helped a great deal. Now I have a working gui. However, using play_url() on a youtube playlist fails though it works with the cli code. Also there's not much in the way of getting the title of the item being casted from these listeners from what I can tell, it just says None for most fields. I was able to get the volume and progress times so it gets the current progress on startup and changing the volume from another device changes the knob in the gui and vice versa. There is still work to be done to support multiple chromecasts correctly but I think I'll upload the code to github once I'm happy with it and share it here.

**Comment by soreau** [comment]
https://github.com/soreau/catt-qt

**Comment by skorokithakis** [comment]
Hey @soreau! That looks cool from the screenshot, but the code doesn't run. You have at least one syntax error (so the file doesn't even load), and you've mixed tabs and spaces so formatters and linters have a hard time with it. I suggest you use Black on the file to make it look consistent.

**Comment by soreau** [comment]
@skorokithakis It runs fine here on Linux and OSX and I didn't use any spaces for indentation. I don't know how to use black. Can you run it through black and submit a patch or tell me how to use it?

**Comment by soreau** [comment]
Will there be a way to play youtube playlists and files from the catt API in the future?

**Comment by theychx** [comment]
Reworking subtitles support and adding some kind of tui is my main priorities right now. But when that is done I will look at it.

**Comment by theychx** [comment]
@soreau ...and your gui looks nice!

**Comment by soreau** [comment]
@thechx Thanks! I hope it runs now.

Also thanks for looking into fixing up the catt api.. it can't exactly cast all the things in it's current state.

**Comment by skorokithakis** [comment]
Thanks @soreau, it looks much better now!

**Comment by soreau** [comment]
Is there a way to recover from a chromecast restart? I tried to re-regiester the cast controller in the status handler when detecting a restart (`if status.app_id == None and status.display_name == None and len(status.namespaces) == 0 and status.session_id == None and status.transport_id == None`) but it fails with
```
ERROR:pychromecast.controllers:Exception thrown when calling cast status listener
Traceback (most recent call last):
  File "/home/user/.local/lib/python3.6/site-packages/pychromecast/socket_client.py", line 1048, in _report_status
    listener.new_cast_status(self.status)
  File "./catt-qt.py", line 368, in new_cast_status
    cast.register_handler(d.device._cast_controller)
  File "/home/user/.local/lib/python3.6/site-packages/pychromecast/socket_client.py", line 375, in register_handler
    self._handlers[handler.namespace] = handler
AttributeError: 'DefaultCastController' object has no attribute 'namespace'
```

**Comment by theychx** [comment]
`pychromecast` also has connection listeners, so I suggest you use that instead:
```
class ConnectionListener:
    def new_connection_status(self, status):
        if status.status == "CONNECTED":
            print("I'm connected!")

connection_listener = ConnectionListener()
cast.register_connection_listener(connection_listener)
```

**Comment by soreau** [comment]
@theychx Thanks for this snippet, I can now see when a chromecast disconnects and reconnects. However, if I try the same routine in response to CONNECTED, I get the same error. I am calling `cast.register_handler(d.device._cast_controller)` and with the existing cast. If I try to create a new cast with `cast = pychromecast.Chromecast(d.device.ip_addr)`, it fails with `pychromecast.error.ChromecastConnectionError: Could not connect to IP:PORT
`

**Comment by soreau** [comment]
I found the problem for reconnecting, which was in pychromecast. I submitted [this pull request](https://github.com/balloob/pychromecast/pull/305) to fix it.

**Comment by soreau** [comment]
Is there a way to unregister listeners? I tried to remove the objects from the respective lists directly but it throws errors such as:
`d.device._cast.media_controller._status_listeners.remove(d.media_listener)
ValueError: list.remove(x): x not in list`

**Comment by soreau** [comment]
[Success](https://github.com/soreau/catt-qt/commit/602b33c8fd7cb5e926c4a72cc5f47a4fe2579f5d) :smile:

**Comment by skorokithakis** [comment]
This looks great! Let us know if you want to propose any changes to the API for things that weren't convenient to use, etc.

I like your GUI a lot, I just think it needs a status display so the user knows what they're waiting for (e.g. when they're starting to play a new video). Also, if you want to package your utility up and publish it at some point, we can link it from our README.

**Comment by soreau** [comment]
@skorokithakis Thanks. I wouldn't have been able to do it without catt and pychromecast. I did notice there is a lack of info() and reboot() functions in the api. Also it cannot play files or youtube playlists (while catt cli can). I recently experienced some play_url() failures on a standard youtube link but I'm hoping these were isolated incidents. Here is one such failure:

```
Traceback (most recent call last):
  File "./catt-qt.py", line 192, in on_play_click
    d.device.play_url(text, resolve=True, block=False)
  File "/home/user/.local/lib/python3.6/site-packages/catt/api.py", line 81, in play_url
    raise APIError("Playback failed")
catt.error.APIError: Playback failed
```

pychromecast could use unregister functions for listeners that have been registered. I was able to remove them from the various lists manually but finding what list took a bit of trudging through the source. This of course doesn't have much to do with catt api.

**Comment by skorokithakis** [comment]
Oh hmm, thanks @soreau, I think an info function would be useful, I don't know about reboot (I thought you couldn't do that remotely?).

I do think these failures are isolated incidents, I get those sometimes too, but pretty rarely. I'm not sure why they happen, I'm afraid. @theychx might know more.

**Comment by soreau** [comment]
@skorokithakis I'm not sure if you saw my [hack](https://github.com/soreau/catt-qt/blob/master/catt-qt.py#L162) to get volume on startup, but it would be nice to remove this and replace it with something like get_info(). Adding reboot functionality would be a nice-to-have for users that do not a capable mobile device but have a computer. I updated status messages such as playback, seek and pause, idle, etc. Let me know if it works ok. So far as packaging is concerned, I am very interested in this. It would be nice to be able to tell users to install it with pip3, but I don't know how to do this yet.

**Comment by soreau** [comment]
@skorokithakis I misread your comment about rebooting slightly, it is certainly possible. See https://github.com/balloob/pychromecast/blob/master/pychromecast/dial.py#L38

**Comment by soreau** [comment]
Ah, turns out I answered my own question. I can call cast.reboot() and it works. It would be nice to include in the catt api.

**Comment by soreau** [comment]
I decided to launch catt instead of calling play_url() from the api. This permits casting files, links and playlist urls as expected. I notice that passing ip is not supported in the current version of catt I have installed with pip3. When might there be a pip3 catt update that has passing ip as device support? Also instead of 'Default Media Reciever' it gets more specific info when playing youtube links with catt at least. I am pleased with the way it's working, feel free to mention in catt readme. I haven't worked out anything in the way of packaging yet but am interested in this so any advice in this direction is welcome. Thanks for your input @skorokithakis and @theychx, it has been a great help.

**Comment by soreau** [comment]
@skorokithakis @theychx catt-qt [is on pypi](https://pypi.org/project/cattqt/) now, you can install it with `pip3 install cattqt`.

**Comment by skorokithakis** [comment]
Thanks @soreau, I'll add it to the README as soon as I can test it out! What do you mean about passing the IP? We support passing an IP as the device already, no?

I do agree that mentioning `reboot()` in the API would be nice, @theychx is there any reason that we shouldn't?

**Comment by soreau** [comment]
@skorokithakis Yes IP support has landed but it's not in latest release as far as I know.

**Comment by soreau** [comment]
@skorokithakis Also it should work on all major platforms and ready to be tested.

**Comment by skorokithakis** [comment]
Oh, I might be running `master` locally and be confused. We should release soon, agreed.

Great news about readiness, I'll test it when I get home and get back to you, thanks!

**Comment by soreau** [comment]
@skorokithakis I noticed at least on windows, it didn't pull in the correct version of pychromecast required for reboot support to work even though I uninstalled and reinstalled catt-qt. You need at least pychromecast 3.2.3 for this functionality.

**Comment by soreau** [comment]
@skorokithakis Now I remember what I was talking about catt having no ip support in released version.. it can be done with the api in release version by passing ip to CattDevice() but the current release catt cli does not have this support, so the ip cannot be used from command line.

**Comment by skorokithakis** [comment]
Oh hmm, I see. So are we all ready to release? @theychx, any objections?

**Comment by theychx** [comment]
This was already included in the `0.10.0` release. :smiley:

**Comment by skorokithakis** [comment]
Bah, I didn't see the release because it wasn't in bold :( Ok, closing, thanks!
